generation/src/proof_generation/proofs/kore.py

```diff
@@ -108,11 +115,44 @@ def kore_exists(var: int) -> Notation:
     return Notation(
         'kore-exists',
         3,
-        _and(sorted_exists(var)(phi0, phi2), App(inhabitant_symbol, phi1)),
+        _and(sorted_exists(var)(phi0, phi2), kore_top(phi1)),
         f'( k∃ {var}:{0} . {2}):{1}',
     )
 
 
+@cache
+def kore_forall(var: int) -> Notation:
+    """kore_forall(inner_sort, outer_sort, pattern)"""
+    return Notation(
+        'kore-forall',
+        3,
+        _and(sorted_forall(var)(phi0, phi2), kore_top(phi1)),
+        f'( k∀ {var}:{0} . {2}):{1}',
+    )
+
+
+def kore_functional(var: int) -> Notation:
+    """kore_functional(var)(sort_of_pattern, outer_sort, pattern)"""
+    return Notation(
+        'kore-functional',
+        3,
+        kore_exists(var)(phi0, phi1, kore_equals(EVar(var), MetaVar(2, e_fresh=(EVar(var),)))),
+        'kore_is_functional({2}:{0}):{1}',
+    )
+
+
+def kore_is_function(arity: list[Pattern]) -> Notation:
+    """kore_is_function(arity)(sort_of_symbol, outer_sort, symbol)"""
+    notation_pattern: Pattern
+    notation_pattern = MetaVar(2, e_fresh=tuple([EVar(x) for x in range(len(arity) + 1)]))
+    for index in range(len(arity)):
+        notation_pattern = App(notation_pattern, EVar(index))
+    notation_pattern = kore_functional(len(arity))(phi0, phi1, notation_pattern)
```

Comment:
```suggestion
    notation_pattern = kore_functional(len(arity) + 1)(phi0, phi1, notation_pattern)
```

since last `x` is `len(arity)`, `y` should start at `len(arity) + 1`